cmd/juju/application/deploy.go

```diff
@@ -496,6 +496,9 @@ type DeployStep interface {
 	// RunPre runs before the call is made to add the charm to the environment.
 	RunPre(MeteredDeployAPI, *httpbakery.Client, *cmd.Context, DeploymentInfo) error
 
+	// RunPreBundle runs before the call is made to add a bundle charm to the environment.
+	RunPreBundle(MeteredDeployAPI, *httpbakery.Client, *cmd.Context, DeploymentInfo) error
```

Comment:
It looks like the purpose of RunPreBundle is to perform charm authorization without checking if the charm is metered. Is that right? 

What is the meaning of obtaining an auth for a non-metered charm, when it's deployed in a bundle?

If we were to leave the check for whether the charm is metered or not, would we be able to get rid of RunPreBundle?

It's just that 
 - the fact that there's no RunPostBundle and 
 - that RunPre calls RunPreBundle
make the DeployStep interface difficult to understand and `deploy.go` is not the lightest read as far as code goes already.

Reply:
I'm ok with attempting to obtain the plan specified in the bundle unconditionally; if we get one from the API, it's reasonable to set it IMO. I agree with @tasdomas though that the method name `RunPreBundle` is confusing.. what does RunPre have to do with bundles? And the logic is getting complicated in there.

I think the general approach in `deployBundle` is fine, but let's please break up the logic and operations in such a way that they can be reused by both the "deploy charm" and "deploy bundle" flow in a clear and concise way.